### PR TITLE
Json

### PR DESCRIPTION
**Pull request by dzove855**
A new Json::to::array is created.

A json like this= { "fuu":  { "luu": "lee" } } will create this:
array[fuu:luu]="lee"

A new json::create
an array like this:
array[fuu:luu]="lee"
will create a json like this:
{ "fuu":  { "luu": "lee" } }

### CODE REVIEW

**Comment by lvenier** [comment]
I think we should start unit test the functions. 
The json functions should always return the same output for the same input. It makes them an ideal case for unit testing. And as they will be very important for any other project we need to make sure we do not break them.

we should add hooks to commit to test the json functions ?
https://sigmoidal.io/automatic-code-quality-checks-with-git-hooks/

Opinion ?

Rgds

File: lib/json.sh [hunk 1]
```diff
@@ -1,73 +1,74 @@
...
-    [[ -z "$1" ]] && return 1
-    [[ -z "$2" ]] && return 2
-
-    typeset -n _json_tmp_array="$2"
-    typeset -n _json_tmp_array_2="$1"
-
-    for key in "${_json_tmp_array[@]}"
-    do
-        _value+="\"$key\","
+Json::create(){
```
[the remark this answers is not in the document]
Comment (dzove855) [reply lib/json.sh]: Because we always start a lib, with the file name, that's why it is Json, but i could create an alias of this function if you want.

**Comment by dzove855** [comment]
I'm for unit testing, but we still did not have implemented something.. so we will wait for the moment..

File: lib/json.sh [hunk 2]
```diff
@@ -1,73 +1,74 @@
...
-    typeset -n array="$arrayname"
-
-    while read -r key
-    do
-        array[$key]="$(echo "$json" | jq -r .$key)"
-    done < <(echo "$json" | jq -r 'keys[]')
+Json::to::array::recursive(){
+    local json="${1}" sub="${2:-keys_unsorted[]}"
+    local parsed_sub="${sub// | keys_unsorted\[\]/}"
+    local parsed_sub="${parsed_sub//keys_unsorted\[\]}"
```
[the remark this answers is not in the document]
Comment (dzove855) [reply lib/json.sh]: Fixed in ned commit

File: lib/json.sh [hunk 3]
```diff
@@ -1,73 +1,74 @@
...
 }
 
-function Json::build::all()
-{
-    Json::create "$_json_array"
+Json::to::array(){
+    # Argmuent should be an array name and a json
+    # the given array will contain the parsed json data
+    
+    local -n array="$1"
```
[the remark this answers is not in the document]
Comment (dzove855) [reply lib/json.sh]: Added a new type check to see if array is associative.

File: lib/json.sh [hunk 4]
```diff
@@ -1,73 +1,74 @@
...
+    local tmpFile="$(mktemp)"
+    local tmpFile2="$(mktemp)"
+    count="0"
+    for entry in "${tmparray[@]}"; do
+        if ! (( count )); then
+            echo "$entry" >> $tmpFile
+        else
+            echo "$entry" > $tmpFile2
+
+            output="$(jq -s -r '.[0] * .[1]' $tmpFile $tmpFile2)"
```
[the remark this answers is not in the document]
Comment (dzove855) [reply lib/json.sh]: Quoted

File: lib/json.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (lvenier) [reply lib/json.sh]: ok 4 alias

**Pull request merged**